lib/java/spring/auth/src/main/java/co/airy/spring/auth/session/CookieSecurityContextRepository.java

```diff
@@ -0,0 +1,117 @@
+ */
+public class CookieSecurityContextRepository implements SecurityContextRepository {
+    private static final Logger log = AiryLoggerFactory.getLogger(CookieSecurityContextRepository.class);
+    private final Jwt jwt;
+
+    public CookieSecurityContextRepository(Jwt jwt) {
+        this.jwt = jwt;
+    }
+
+    @Override
+    public SecurityContext loadContext(HttpRequestResponseHolder requestResponseHolder) {
+        HttpServletRequest request = requestResponseHolder.getRequest();
+        HttpServletResponse response = requestResponseHolder.getResponse();
+        requestResponseHolder.setResponse(new SaveToCookieResponseWrapper(request, response));
+
+        SecurityContext context = SecurityContextHolder.createEmptyContext();
+        getStoredAuth(request)
+                .ifPresent(context::setAuthentication);
+
+        return context;
+    }
+
+    @Override
+    public void saveContext(SecurityContext context, HttpServletRequest request, HttpServletResponse response) {
+        SaveToCookieResponseWrapper responseWrapper = WebUtils.getNativeResponse(response,
+                SaveToCookieResponseWrapper.class);
+        if (responseWrapper != null && !responseWrapper.isContextSaved()) {
+            responseWrapper.saveContext(context);
+        }
+    }
+
+    @Override
+    public boolean containsContext(HttpServletRequest request) {
+        return getStoredAuth(request).isPresent();
+    }
+
+    private Optional<AiryAuth> getStoredAuth(HttpServletRequest request) {
+        return getCookie(request)
+                .map((authCookie) -> {
+                    try {
+                        return jwt.loadFromToken(authCookie.getValue());
+                    } catch (Exception e) {
+                        log.warn("Clearing user session because token could not be decoded", e);
+                        return null;
+                    }
+                });
+    }
+
+    private Optional<Cookie> getCookie(HttpServletRequest request) {
+        if (request.getCookies() == null) {
+            return Optional.empty();
+        }
+
+        return Stream.of(request.getCookies())
+                .filter(c -> AuthCookie.NAME.equals(c.getName()) && !c.getValue().equals(""))
+                .findFirst();
+    }
+
+    private class SaveToCookieResponseWrapper extends SaveContextOnUpdateOrErrorResponseWrapper {
+        private final HttpServletRequest request;
+
+        SaveToCookieResponseWrapper(HttpServletRequest request, HttpServletResponse response) {
+            super(response, true);
+            this.request = request;
+        }
+
+        @Override
+        protected void saveContext(SecurityContext securityContext) {
+            HttpServletResponse response = (HttpServletResponse) getResponse();
+            Authentication authentication = securityContext.getAuthentication();
+
+            if (authentication instanceof OAuth2AuthenticationToken) {
```

Comment:
See the class comment.

If the current security context is set to a successful oauth2 authentication then we set an httpOnly, secure jwt cookie. This was necessary because it's unfeasible to store the entire oauth2 context in a client-side jwt.

Implications:
- All subsequent requests will be authenticated with this token
- Clients can clear this cookie by visiting `/logout`
- Prevents session hijacking through Javascript based attacks
- the original oauth2 context gets lost. So the `ProfileData.java` class has to include all user information needed for logging, workflow etc.